### ISSUE TITLE
Socket time-outs cause "ghost" connection on a reconnect.

### ISSUE THREAD

**Issue by Neshfi** [nkzawa/socket.io-client.java#87]
Ok, my title may not be very clear, but it's also a weird problem, so please bare with me for a moment. I will try to explain everything as clear as possible.

### The issue
The issue that I got is that I'm dealing with ghost connection when a device loses connection for a while and then reconnects again after several timeouts. The ghost connections aren' t a big deal, cause they do time-out after a while, but they are annoying and should't even exists none the less.

### The client and server code
**Client**
```java
    private void openSocket() throws URISyntaxException {
        IO.Options opts = new IO.Options();
        opts.reconnection = true;
        opts.reconnectionDelay = 1000;
        opts.timeout = 10000;
        opts.query = "uid=" + Build.SERIAL;

        socket = IO.socket(URL, opts);

        socket.on(Socket.EVENT_CONNECT, onConnect);
        socket.on(Socket.EVENT_CONNECT_TIMEOUT, onConnectionTimeout);
        socket.on(Socket.EVENT_DISCONNECT, onDisconnect);
        socket.on(Socket.EVENT_ERROR, onError);

        socket.connect();
    }
```
(All event handlers do nothing but log stuff)

**Server**
```javascript
var app = require('express')();
var http = require('http').Server(app);
var io = require('socket.io')(http);

io.on('connection', function(socket){
	var handshakeQuery = socket.handshake.query;
	console.log("Android device " + handshakeQuery.uid + " connected at " + new Date().timeNow());

	socket.on('disconnect', function(msg){
		console.log("Android device " + handshakeQuery.uid + " disconnected");
	});
});

http.listen(3001, function(){
	console.log('listening on *:3001');
});
```
### How to reproduce
1. Start the server.
2. Start the Android client.
3. Close the server.
4. Wait for the client to time-out 1 or more times.
5. Start the server again.
You should see several connections popping up in the server console. All from the same Android device.

### My results from different scenario's
**Scenario 1 - Normal connection**
I start the server and I start the client. 

Client logs:
```
3069-3096/com.example.myapp I/SocketManager﹕ Socket connected
```
Server logs:
```
listening on *:3001
Android device GDX1ZDF5KR connected at 10:02:19
```
Everything goes fine without a problem

**Scenario 2 - Reconnect before time-out**
I start the server and I start the client, then I close the server and start it back up again before the client times-out. 

Client logs:
```
3069-3096/com.example.myapp I/SocketManager﹕ Socket connected
3069-3101/com.example.myapp I/SocketManager﹕ Socket disconnected
3069-3115/com.example.myapp I/SocketManager﹕ Socket connected
```
Server logs:
```
listening on *:3001
Android device GDX1ZDF5KR connected at 10:02:19
^C
C:\wamp\www\socket-server-example>node .
listening on *:3001
Android device GDX1ZDF5KR connected at 10:02:30
```
Everything still goes fine without a problem. The client connects to the server, disconnects because the server gets closed and connects again when the server is back on. So far no ghost connections.

**Scenario 3 - Reconnect after one time-out**
I start the server and I start the client, then I close the server and wait for the client to time-out once. After that I start the server again.

Client logs:
```
3069-3115/com.example.myapp I/SocketManager﹕ Socket connected
3069-3116/com.example.myapp I/SocketManager﹕ Socket disconnected
3069-3121/com.example.myapp I/SocketManager﹕ Connection timed out
3069-3132/com.example.myapp I/SocketManager﹕ Socket connected
```
Server logs:
```
listening on *:3001
Android device GDX1ZDF5KR connected at 10:02:30
^C
C:\wamp\www\socket-server-example>node .
listening on *:3001
Android device GDX1ZDF5KR connected at 10:03:04
Android device GDX1ZDF5KR connected at 10:03:13
```
Now this is where is starts to go wrong. In the Android logs I only see it reconnect once, while in the serverlogs I see it connect twice. The connection that got started at 10:03:04 is usable for communication between the server and client. The connection that started on 10:03:13 is useless for communication as it is not bound to any events on the client side. It also times out after the time-out interval that is set on the client side.

**Scenario 4 - Reconnect after three time-out**
I start the server and I start the client, then I close the server and wait for the client to time-out three times. After that I start the server again.

Client logs:
```
3069-3132/com.example.myapp I/SocketManager﹕ Socket connected
3069-3135/com.example.myapp I/SocketManager﹕ Socket disconnected
3069-3140/com.example.myapp I/SocketManager﹕ Connection timed out
3069-3143/com.example.myapp I/SocketManager﹕ Connection timed out
3069-3146/com.example.myapp I/SocketManager﹕ Connection timed out
3069-3158/com.example.myapp I/SocketManager﹕ Socket connected
```
Server logs:
```
listening on *:3001
Android device GDX1ZDF5KR connected at 10:03:20
^C
C:\wamp\www\socket-server-example>node .
listening on *:3001
Android device GDX1ZDF5KR connected at 10:04:00
Android device GDX1ZDF5KR connected at 10:04:06
Android device GDX1ZDF5KR connected at 10:04:07
Android device GDX1ZDF5KR connected at 10:04:18
```
As you can see, it went from bad in scenario 3 to worse. The connection at 10:04:00 is a usable connection, all other connections after that are all ghost connections. This is also the scenario in which I thing that the amount of ghost connections has a relation with the amount of time-outs on the client. In scenario 3 I timed-out once and only got one ghost connection. In this scenario I timed-out three times and also got three ghost connections. I have done some more experiments to verify this and it seems like it holds.

Sorry for the long post, but I hoped that I gave you enough information to work with.

**Comment by samarthagrawal** [comment]
Hi Neshfi, How did you solve this problem eventually ?

**Comment by nkzawa** [comment]
Sorry for the delay and Thank you very much for you great explanations!
I'm working on the issue.

**Comment by Neshfi** [comment]
@samarthagrawal I haven't been able to solve it yet.

@nkzawa Thank you for looking into it. I noticed that when I had server offline for offline for a very long time (Like 15 hours or so), then I had no problems at all. I don't know if this little bit of information helps, but I guess I share it with you anyway.

**Comment by nkzawa** [comment]
Umm, I tried to reproduce the issue but couldn't somehow.
As far as I tried, `EVENT_CONNECT_ERROR` event happens first, so that `EVENT_CONNECT_TIMEOUT` never does. 

The socket options I used for the client is the following:
```js
IO.Options opts = new IO.Options();
opts.reconnection = true;
opts.reconnectionDelay = 1000;
opts.timeout = 10000;
opts.query = "uid=foo";
```

Additionally, I changed the `timeout` option to `1000`, so that `EVENT_CONNECT_TIMEOUT` happens, but anyway ghost connections didn't occur.

I tested this on my local machine, not on Android. The lib version I used is v0.4.0 (latest).

**Comment by Neshfi** [comment]
My project used 0.3.0 and I used it on an Android device. I will try it again tomorrow (Since all my stuff is at work) and see if I still got that problem with 0.4.0. 

When you say that you tested it on your local machine, I guess you used an emulator? If so, then I will create an emulator enviroment aswell and see if it gives the same result or not.

**Comment by nkzawa** [comment]
Thank you for your reply. I tested this on local Java environment, not on an Android emulator.

**Comment by nkzawa** [comment]
I tested this with v0.3.0 but the result is the same. Maybe this happens only on Android.

**Comment by samarthagrawal** [comment]
Hi Naoyuki, 

I am not sure if it is related but ghost connections also occur on EVENT_ERROR when Android side socket.io tries to reconnect automatically and the server requires an authentication cookie. In this case a new socket is connected and the old socket stays hanging on the server side (which disconnects after the set time). My guess is that after the connection breaks, the Android side socket.io tries to do another handshake, goes through the authentication middleware at the server which forces a new connection. 

Regards,
Samarth Agrawal

**Comment by Neshfi** [comment]
A day too late, sorry for that ... but the problem is still there. I've updated to 0.4.0 and I've used different kinds of Android devices, but they all cause ghost connections on time-outs.

**Comment by nkzawa** [comment]
I think found it. This problem is because a http request sometimes stop and abort after about 63 seconds (on my Android device) if it couldn't connect to sever immediately.

As far as I researched there is no way to abort a connection manually, so I just set a connect timeout for fixing the issue.

https://github.com/nkzawa/engine.io-client.java/commit/2aac8b65c15722d37a77ddf897bcc1adcd881bf9

Still one or two ghost connections might sometimes happen, but in the most cases,  it'd be no problem.
